### PR TITLE
[GO][Client] Return error in Execute() methods

### LINKED ISSUE #8263: [BUG][GO][Client] Returned err is never nil
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?

##### Description

When calling any generated `Execute()` function, a non-nil error is always returned. Instead of using the usual Go idiom:

```go
result, _, err := r.Execute()
if err != nil {
	/* deal with error */
}
```

I need to check the actual error message in order to detect errors:

```go
result, _, err := r.Execute()
if err != nil && err.Error() != "" {
	/* deal with error */
}
```

##### openapi-generator version

c3220848 (current main branch)

##### Related issues/PRs

I believe this was introduced by #8137. Maybe it is intended behaviour. Care to weigh in here, @code-lucidal58?

##### Suggest a fix

Successful execution of `*Execute()` should not return a non-nil error.

### PR DESCRIPTION
**Pull request by thiagoarrais**
Since #8137 the generated Go client no longer returns `error` results. It instead prefers the openapi-generator specific `GenericOpenAPIError` concrete type. This breaks expectations from Go developers: the main one being that returned errors are nullable and that a null error result means successful conclusion.

This PR reverts that change but tries to keep up with the spirit of the original PR by still allowing the response body to be accessed by client code. It also changes the generated documentation to make the possibility of getting detailed errors clearer.

Fixes #8263 (please refer to issue for aditional discussion).

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

cc @antihax (2017/11) @grokify (2018/07) @kemokemo (2018/09) @bkabrda (2019/07)

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/go/api_doc.mustache [hunk 1]
```diff
@@ -35,6 +35,11 @@ import (
     {{goImportAlias}} "./openapi"
 )
 
+type openapiError interface {
+    Model() interface{}
+    Body() []byte
+}
+
```
Comment (zippolyte) [inline modules/openapi-generator/src/main/resources/go/api_doc.mustache:42]: Not sure I understand the need for a custom interface. You can directly do `err.(GenericOpenAPIError)`

File: modules/openapi-generator/src/main/resources/go/api_doc.mustache [hunk 2]
```diff
@@ -46,6 +51,10 @@ func main() {
     if err != nil {
         fmt.Fprintf(os.Stderr, "Error when calling `{{classname}}.{{operationId}}``: %v\n", err)
         fmt.Fprintf(os.Stderr, "Full HTTP response: %v\n", r)
+        if openapiErr, ok := err.(openapiError); ok {
+            fmt.Fprintf(os.Stderr, "Model returned from error response: %v\n", openapiErr.Model())
+            fmt.Fprintf(os.Stderr, "Raw response body: %v\n", openapiErr.Body())
```
Comment (zippolyte) [inline modules/openapi-generator/src/main/resources/go/api_doc.mustache:56]:
```suggestion
            fmt.Fprintf(os.Stderr, "Raw response body: %s\n", openapiErr.Body())
```

File: modules/openapi-generator/src/main/resources/go/api_doc.mustache (hunk 1 above)
Comment (thiagoarrais) [reply modules/openapi-generator/src/main/resources/go/api_doc.mustache:42]: Yup. Will do that.

**Review by zippolyte — APPROVED** [review]
This LGTM !
Thanks

**Comment by wing328** [comment]
Please put this work on hold for the time being.

Likely we will file another PR to correct the behaviour.

**Comment by thiagoarrais** [comment]
Closing in favour of #8427 + #8439

**Pull request closed without merging**